### PR TITLE
[1/N] Rs/vllm quantization - Refactor to minimize `llama.py` changes

### PR DESCRIPTION
**Pull request by robertgshaw2-neuralmagic**
Paired with @dsikka to refactor `SmoothQuantLinearMethod` to avoid making changes to `llama.py`
- Removed all the "layer specific" `SmoothQuantLinearMethod` by making the indexing (splitting QKV into logical shards generic and explicitly handling state_dict converion
- Successfully whittled down to only add one LOC to `llama.py`

Many todos left, including:
- We currently have hardcoded `use_per_token`, need to use the quant config for this
- We need a way to pass different quantconfigs to each layer to support nonuniform quantization

### CODE REVIEW

File: vllm/model_executor/layers/linear.py [hunk 1]
```diff
@@ -59,7 +65,7 @@ def __init__(self, separate_bias_add: bool = False):
     def create_weights(self, input_size_per_partition: int,
                        output_size_per_partition: int, input_size: int,
                        output_size: int,
-                       params_dtype: torch.dtype) -> Dict[str, Any]:
+                       params_dtype: torch.dtype, logical_widths: Optional[List[int]]) -> Dict[str, Any]:
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/linear.py:68]: lift this to be inside `LinearMethodBase` ?

File: vllm/model_executor/layers/quantization/smoothquant.py [hunk 2]
```diff
@@ -1,8 +1,9 @@
-from typing import Any, Dict, List, Tuple, Optional
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/quantization/smoothquant.py:1]: very nice cleanup 🙌

File: vllm/model_executor/model_loader.py [hunk 3]
```diff
@@ -83,8 +83,7 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
         # The weights will be initialized as empty tensors.
         with torch.device(device_config.device):
             if _is_support_smoothquant(model_config):
-                model = model_class(model_config.hf_config, linear_method,
-                                    quant_config)
+                model = model_class(model_config.hf_config, linear_method)
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/model_loader.py:86]: How come we don't have to pass in the `quant_config` ?

**Comment by varun-sundar-rabindranath** [comment]
LGTM.

**Review by varun-sundar-rabindranath — APPROVED** [review]

File: vllm/model_executor/layers/linear.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:68]: I got rid of this on the next pr

File: vllm/model_executor/model_loader.py (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/model_loader.py:86]: Yeah linear method handles it

**Pull request merged**